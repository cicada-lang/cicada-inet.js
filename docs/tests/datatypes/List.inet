type Trivial 0 end
node sole value!: Trivial end

type List 1 end
node null value!: List('A) end
node cons head: 'A tail: List('A) -- value!: List('A) end

node append target!: List('A) rest: List('A) -- return: List('A) end

rule null append
  (append)-rest
  return-(append)
end

rule cons append
  (append)-rest (cons)-tail append
  (cons)-head cons
  return-(append)
end

claim six_soles Trivial List end

define six_soles
  null sole cons sole cons sole cons
  null sole cons sole cons sole cons
  append
end

show six_soles end
run six_soles end