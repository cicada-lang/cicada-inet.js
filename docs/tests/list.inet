(define-cons sole
  (-> ()
      (Trivial)))

(define-cons null
  (-> ()
      (TrivialList)))

(define-cons cons
  (-> (TrivialList Trivial)
      (Nat)))

(define-elim append
  (-> (TrivialList TrivialList)
      (TrivialList)))

(define-rule
  (null append)
  ())

(define-rule
  (cons append)
  (rot rot append swap cons))

(define-net six-soles
  null sole cons sole cons sole cons
  null sole cons sole cons sole cons
  append)
