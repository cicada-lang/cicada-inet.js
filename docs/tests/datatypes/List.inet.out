(cons₅)-value!target-(append₀)
(cons₂)-value rest-(append₀)
(cons₄)-value tail-(cons₅)
(sole₅)-value head-(cons₅)
(cons₃)-value tail-(cons₄)
(sole₄)-value head-(cons₄)
(null₁)-value tail-(cons₃)
(sole₃)-value head-(cons₃)
(cons₁)-value tail-(cons₂)
(sole₂)-value head-(cons₂)
(cons₀)-value tail-(cons₁)
(sole₁)-value head-(cons₁)
(null₀)-value tail-(cons₀)
(sole₀)-value head-(cons₀)
(append₀)-return

(cons₂)-value tail-(cons₈)
(cons₈)-value tail-(cons₇)
(sole₃)-value head-(cons₈)
(cons₇)-value tail-(cons₆)
(sole₄)-value head-(cons₇)
(sole₅)-value head-(cons₆)
(cons₁)-value tail-(cons₂)
(sole₂)-value head-(cons₂)
(cons₀)-value tail-(cons₁)
(sole₁)-value head-(cons₁)
(null₀)-value tail-(cons₀)
(sole₀)-value head-(cons₀)
(cons₆)-value

